homeassistant/components/recorder/migration.py

```diff
@@ -689,13 +687,106 @@ def _restore_foreign_key_constraints(
+            )
+            _delete_foreign_key_violations(
+                session_maker, table, column, foreign_table, foreign_column
+            )
+            _add_constraint(session_maker, add_constraint, table, column)
+
+
+def _add_constraint(
+    session_maker: Callable[[], Session],
+    add_constraint: AddConstraint,
+    table: str,
+    column: str,
+) -> None:
+    """Add a foreign key constraint."""
+    _LOGGER.warning(
+        "Adding foreign key constraint to %s.%s. "
+        "Note: this can take several minutes on large databases and slow "
+        "machines. Please be patient!",
+        table,
+        column,
+    )
+    with session_scope(session=session_maker()) as session:
+        try:
+            connection = session.connection()
+            connection.execute(add_constraint)
+        except (InternalError, OperationalError):
+            _LOGGER.exception("Could not update foreign options in %s table", table)
+
+
+def _delete_foreign_key_violations(
+    session_maker: Callable[[], Session],
+    table: str,
+    column: str,
+    foreign_table: str | None,
+    foreign_column: str | None,
+) -> None:
+    """Remove rows which violate the constraints."""
+    if foreign_table is None:
+        return
+
+    _LOGGER.warning(
+        "Rows in table %s where %s references non existing %s.%s will be %s. "
+        "Note: this can take several minutes on large databases and slow "
+        "machines. Please be patient!",
+        table,
+        column,
+        foreign_table,
+        foreign_column,
+        "set to NULL" if table == foreign_table else "deleted",
+    )
 
+    if table == foreign_table:
+        # In case of a foreign reference to the same table, we set invalid
+        # references to NULL instead of deleting as deleting rows may
+        # cause additional invalid references to be created. This is to handle
+        # old_state_id referencing a missing state.
         with session_scope(session=session_maker()) as session:
-            try:
-                connection = session.connection()
-                connection.execute(add_constraint)
-            except (InternalError, OperationalError):
-                _LOGGER.exception("Could not update foreign options in %s table", table)
+            # The subquery (SELECT {foreign_column} from {foreign_table}) is
+            # to be compatible with old MySQL versions which do not allow
+            # referencing the table being updated in the WHERE clause.
+            session.execute(
+                text(
+                    f"UPDATE {table} "  # noqa: S608
+                    f"SET {column} = NULL"
+                    " WHERE ("
+                    f" {table}.{column} IS NOT NULL AND"
+                    "  NOT EXISTS"
+                    "  (SELECT 1 "
+                    f"   FROM  (SELECT {foreign_column} from {foreign_table}) AS t2"
+                    f"   WHERE t2.{foreign_column} = {table}.{column}));"
+                )
```

Comment:
Do we need to do these in batches of 100000 to avoid running out of row locks like we do with the other mass updates?